air/src/execution_step/value_types/canon_stream_map.rs

```diff
@@ -0,0 +1,291 @@
+            map.entry(key)
+                .and_modify(|canon_stream| {
+                    canon_stream.push(&value);
+                })
+                .or_insert(CanonStream::from_values_and_tetraplet(
+                    vec![value.clone()],
+                    canon_stream.tetraplet().clone(),
+                ));
+        }
+        let values = canon_stream.get_values().clone();
+        Ok(Self { values, map, tetraplet })
+    }
+
+    // This returns a number of distinct keys in the canon map.
+    pub(crate) fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    pub(crate) fn as_jvalue(&self) -> JValue {
+        // TODO: this clone will be removed after boxed values
+        let jvalue_array = self
+            .values
+            .iter()
+            .map(|r| r.get_result().deref().clone())
+            .collect::<Vec<_>>();
+        JValue::Array(jvalue_array)
+    }
+
+    pub(crate) fn iter(&self) -> impl ExactSizeIterator<Item = &ValueAggregate> {
+        self.values.iter()
+    }
+
+    pub(crate) fn tetraplet(&self) -> &Rc<SecurityTetraplet> {
+        &self.tetraplet
+    }
+
+    pub(crate) fn index<'self_l>(
+        &'self_l self,
+        stream_map_key: &StreamMapKey<'key>,
+    ) -> ExecutionResult<&'self_l CanonStream> {
+        println!("index map {:#?} ", self.map);
+        let canon_stream = self
+            .map
+            .get(stream_map_key)
+            .ok_or(UncatchableError::CanonStreamMapError(IndexIsAbsentInTheMap))?;
+        println!("index canon {:#?} ", canon_stream);
+
+        Ok(canon_stream)
+    }
+}
+
+use std::fmt;
+
+impl fmt::Display for CanonStreamMap<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[")?;
+        for (key, canon_stream) in self.map.iter() {
+            write!(f, "{key:?} : {canon_stream:?}, ")?;
+        }
+        write!(f, "]")
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CanonStreamMapWithProvenance<'a> {
+    pub(crate) canon_stream_map: CanonStreamMap<'a>,
+    pub(crate) cid: Rc<CID<CanonResultCidAggregate>>,
+}
+
+impl<'a> CanonStreamMapWithProvenance<'a> {
+    pub(crate) fn new(canon_stream_map: CanonStreamMap<'a>, cid: Rc<CID<CanonResultCidAggregate>>) -> Self {
+        Self { canon_stream_map, cid }
+    }
+}
+
+impl<'a> Deref for CanonStreamMapWithProvenance<'a> {
+    type Target = CanonStreamMap<'a>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.canon_stream_map
+    }
+}
+
+fn get_value_from_obj(value_aggregate: &ValueAggregate) -> ExecutionResult<ValueAggregate> {
+    let tetraplet = value_aggregate.get_tetraplet();
+    let provenance = value_aggregate.get_provenance();
+    let trace_pos = value_aggregate.get_trace_pos();
+    let object = value_aggregate
+        .get_result()
+        .as_object()
+        .ok_or(UncatchableError::StreamMapKeyError(NotAnObject))?;
```

Comment:
I would suggest importing `StreamMapKeyError` variants only within this function.

Reply:
Done